### PR TITLE
Create reusable configure and build action

### PR DESCRIPTION
**Pull request by amgebauer**
## Description and Context

Create composite actions to reuse configure and build 4C.

### CODE REVIEW

**Review by sebproell — APPROVED** [review]
LGTM, but I'll let @ppraegla decide if this makes sense in combination with the planned next steps.

**Comment by amgebauer** [comment]
#5 will also add a similar action for build. I will combine these two into this PR.

**Comment by amgebauer** [comment]
With the latest changes, it's now possible to reuse configure and build in other workflows. I did it in #5 for doxygen and readthedocs.

Not, this PR is ready for review.

File: .github/workflows/buildtest.yml
Comment (amgebauer) [inline .github/workflows/buildtest.yml:53]: @ppraegla This was missing. I added it on the way.

File: .gitlab/CODEOWNERS
Comment (ppraegla) [inline .gitlab/CODEOWNERS:418]:
```suggestion
.github/                                                             @baci/team_infrastructure
```
I think we could just name the `.github/` folder to save us the trouble to adapt in other PRs. The Gitlab codeowners file does not make sense anyway once we are fully on Github.

File: .github/actions/build_4C/action.yml
Comment (ppraegla) [reply .github/actions/build_4C/action.yml:20]: I think it is fine because when we set the shell to bash below. The option `pipefail` is added automatically. 
See `shell: bash --noprofile --norc -e -o pipefail {0}` from the workflow log.

But I agree that we might not even need this log. But we can keep it for now and remove it later.

File: .github/workflows/buildtest.yml
Comment (ppraegla) [inline .github/workflows/buildtest.yml:43]: I'm a bit confused whats happening here (not your code. But what the github runner is doing). You set the build directory to `${{ github.workspace }}/../build` which according to the [log](https://github.com/4C-multiphysics/4C/actions/runs/11763968417/job/32768579260#step:6:27) is `BUILD_DIRECTORY: /home/runner/work/4C/4C/../build`. At the end of the configuring the [log](https://github.com/4C-multiphysics/4C/actions/runs/11763968417/job/32768579260#step:6:522) says `Build files have been written to: /__w/4C/build`. The subsequent build works though. So it seems that`$GITHUB_WORKSPACE` and `${{ github.workspace }}` are after all the same directory.

**Review by ppraegla — APPROVED** [review]
Nice work. Only some minor comments. You can merge if you want.

**Review by sebproell — APPROVED** [review]

File: .github/actions/build_4C/action.yml
Comment (amgebauer) [reply .github/actions/build_4C/action.yml:20]: Let me remove it. We don't even need it for readthedocs.

File: .github/workflows/buildtest.yml
Comment (amgebauer) [reply .github/workflows/buildtest.yml:43]: Yes, that also confused me.

Comment (sebproell) [reply .github/workflows/buildtest.yml:43]: Slightly related: how about using a subfolder `build` inside the repo in the future? Actually, I think we should even ignore this folder in the `.gitignore` since this is how I nowadays recommend setting up a build.

Comment (amgebauer) [reply .github/workflows/buildtest.yml:43]: good idea. I do that for a few other jobs to avoid unnecessary copying and avoiding relative links for artifacts. Only for clang-tidy, it might not work (or we might need to exclude the build-folder then)

Comment (sebproell) [reply .github/workflows/buildtest.yml]: Why should it not work for clang-tidy? We are normally using the `-p` arg which tells clang-tidy where to look for compile command lines. I use that locally all the time.

File: .gitlab/CODEOWNERS
Comment (amgebauer) [reply .gitlab/CODEOWNERS:418]: I left it for now like so as we also have the PR/Issue templates. We can change it later if we decide how we will proceed with codeowners.

File: .github/workflows/buildtest.yml
Comment (amgebauer) [reply .github/workflows/buildtest.yml]: We then also get warnings/errors from our dependencies (ArborX, Google test, etc).

Comment (sebproell) [reply .github/workflows/buildtest.yml]: Ah I see, it is the combination of `-p compile_commands.json` and the additional working directory. I am fine with leaving it as is.

**Review by sebproell — APPROVED** [review]

**Comment by amgebauer** [comment]
I had to fix a path. The tests passed, but I need another approval.

**Review by sebproell — APPROVED** [review]

**Pull request merged**
